pyro/distributions/transforms/generalized_channel_permute.py

```diff
@@ -2,38 +2,131 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import torch
-import torch.nn.functional as F
-from torch.distributions import constraints
+from torch.distributions import Transform, constraints
 
+from pyro.distributions.conditional import ConditionalTransformModule
 from pyro.distributions.torch_transform import TransformModule
 from pyro.distributions.util import copy_docs_from
+from pyro.nn import DenseNN
 
 
-@copy_docs_from(TransformModule)
-class GeneralizedChannelPermute(TransformModule):
-    """
+@copy_docs_from(Transform)
+class ConditionedGeneralizedChannelPermute(Transform):
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 3
+
+    def __init__(self, permutation=None, LU=None):
+        super(ConditionedGeneralizedChannelPermute, self).__init__(cache_size=1)
+
+        self.permutation = permutation
+        self.LU = LU
+
+    @property
+    def U_diag(self):
+        return self.LU.diag()
+
+    @property
+    def L(self):
+        return self.LU.tril(diagonal=-1) + torch.eye(self.LU.size(-1), dtype=self.LU.dtype, device=self.LU.device)
+
+    @property
+    def U(self):
+        return self.LU.triu()
+
+    def _call(self, x):
+        """
+        :param x: the input into the bijection
+        :type x: torch.Tensor
+
+        Invokes the bijection x=>y; in the prototypical context of a
+        :class:`~pyro.distributions.TransformedDistribution` `x` is a sample from
+        the base distribution (or the output of a previous transform)
+        """
+
+        """
+        NOTE: As is the case for other conditional transforms, the batch dim of the
+        context variable (reflected in the initial dimensions of filters in this
+        case), if this is a conditional transform, must broadcast over the batch dim
+        of the input variable.
+
+        Also, the reason the following line uses matrix multiplication rather than
+        F.conv2d is so we can perform multiple convolutions when the filters
+        "kernel" has batch dimensions
+        """
+        filters = (self.permutation @ self.L @ self.U)[..., None, None]
+        y = (filters * x.unsqueeze(-4)).sum(-3)
```

Comment:
are you sure `einsum` isn't preferable here?

Reply:
Yes, I find it's harder to understand as well as being slower